sdk/core/azure-core/azure/core/pipeline/policies/_utils.py

```diff
@@ -102,3 +102,19 @@ def get_domain(url: str) -> str:
     :return: The domain of the url.
     """
     return str(urlparse(url).netloc).lower()
+
+
+def sanitize_url_query_params(url: str, allowed_query_params: Set[str], placeholder: str = "REDACTED") -> str:
+    """Redact query parameters from the URL.
+
+    :param str url: The URL to sanitize.
+    :param list[str] allowed_query_params: The list of allowed query parameters.
+    :param str placeholder: The placeholder to use for redacted query parameters.
+    :rtype: str
+    :return: The sanitized URL.
+    """
+    parsed_url = list(urlparse(url))
+    parsed_qp = parse_qsl(parsed_url[4], keep_blank_values=True)
+    filtered_qp = [(key, value if key.lower() in allowed_query_params else placeholder) for key, value in parsed_qp]
```

Comment:
Not sure if we should assume that query parameters are case insensitive...

Reply:
For this, I am following the pattern already established in the query param sanitization in the HttpLoggingPolicy. I believe other languages also do case-insensitive comparison when comparing query params with values in the allow list ([example](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Shared/HttpMessageSanitizer.cs#L116).

I think for the sake of allow-listing for sanitization, case-insensitivity should be fine. So if someone passes in `allowed_query_params=["Comp", "res"]`, it should be practical to allow `comp`. The `allowed_query_params` passed in to this method are already expected to be lowercased.

Reply:
From a practical perspective, I don't know how much of a difference it makes. But it is technically not correct to treat them as case insensitive (as opposed to headers, which in HTTP/1.1 *are* case insensitive).

What is our recommended usage for users of the library? And what is the recommended usage for client developers? And should we use the same allow-list as the logging policy?

Reply:
I feel like if something is safe to log, then it is safe to keep in traces, so it would be nice to allow some client-level parameter that could apply to both `HttpLoggingPolicy` and `DistributedTracingPolicy`. It's definitely doable with policy constructor parameters.

What I imagine usage to be like:

1. **SDK users:** If some query parameter is redacted in their traces that they'd like allowed, they can pass it in via some keyword argument in their service client that DistributedTracingPolicy will honor.
   - `client = ServiceClient(credential, allowed_query_params={'foo'})`
3. **Client Developers**: If they want to have some default list of allowed query params, then they can either employ a similar approach to some existing SDKs modifying the allowlist for HttpLoggingPolicy:
`http_logging_policy.allowed_query_params.update({...})` ([example](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/formrecognizer/azure-ai-formrecognizer/azure/ai/formrecognizer/_form_base_client.py#L92)).
   - Client developers could also update any user-provided `allowed_query_params` in the kwargs with their defaults, so they don't have to get references to the policies and can just propagate the kwargs.

Also, I am thinking we should make query parameters passed in via this kwarg additive by doing a set union vs replacing the set. This is how Java [treats](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/UrlSanitizer.java#L40) user-provided parameters. JS uses an additive approach as well and calls it [`additionalAllowedQueryParameters`](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/core/core-rest-pipeline/src/policies/tracingPolicy.ts#L38).


---

Summary, what do you think about:
   - Having both `HttpLoggingPolicy` and `DistributedTracingPolicy` support kwarg `allowed_query_params`.
   - The `allowed_query_params` kwarg will be additive with the default allow list (which will only contain "api-version" right now)
   -  `DistributedTracingPolicy` will introduce `self.allowed_query_params` as an instance variable similar to `HttpLoggingPolicy`.